### ISSUE TITLE
lsp does not identify the correct project root in a subfolder

### ISSUE THREAD

**Issue by rubenmoor** [haskell/haskell-language-server#2227]
### My environment

```sh
haskell-language-server version: 1.2.0.0 (GHC: 8.10.7) (PATH: /nix/store/ay3hgsg6df98avgj8lywrdg32bi8yljm-haskell-language-server-1.2.0.0/bin/haskell-language-server)
Tool versions found on the $PATH
cabal:		3.4.0.0
stack:		2.7.3
ghc:		8.10.7
```
On NixOS, emacs lsp

Contents of cabal file
```
cabal-version:      2.4
name:               static-css
version:            0.1.0.0
author:             Ruben Moor
maintainer:         ruben.moor@gmail.com
extra-source-files: CHANGELOG.md
executable static-css
    main-is:          Main.hs
    build-depends:    base
                    , clay
    hs-source-dirs:   app
    default-language: Haskell2010
```

Contents of `hie.yaml`:
```yaml
cradle:
  cabal:
    - path: "./app"
      component: "exe:static-css"
```

Now this project is in a subfolder of a bigger haskell project that has its own hie.yaml.

When I run ```hie-bios debug app/Main.hs``` I get

```
Root directory:        /home/ruben/code/learn-palantype/static-css
Component directory:   /home/ruben/code/learn-palantype/static-css
...
```

and everything seems fine. In my emacs, however, lsp exits and the stderr-buffer reads

```
Cradle directory: /home/ruben/code/learn-palantype
```

which is clearly wrong.

**Comment by jneira** [comment]
Hi, thanks for the bug report. How do you open the project in emacs? directly over the `/home/ruben/code/learn-palantype/static-css` folder? if you have the chance of checking it, does it fail in vscode?

//cc @fendor

**Comment by fendor** [comment]
Hello!
Can you show the output of `haskell-language-server --debug app/Main.hs`?

**Comment by rubenmoor** [comment]
> Hello!
> Can you show the output of `haskell-language-server --debug app/Main.hs`?

```shell
$ cd /home/ruben/code/learn-palantype/static-css
$ haskell-language-server --debug app/Main.hs

haskell-language-server version: 1.2.0.0 (GHC: 8.10.7) (PATH: /nix/store/ay3hgsg6df98avgj8lywrdg32bi8yljm-haskell-language-server-1.2.0.0/bin/haskell-language-server)
 ghcide setup tester in /home/ruben/code/learn-palantype/static-css.
Report bugs at https://github.com/haskell/haskell-language-server/issues

Step 1/4: Finding files to test in /home/ruben/code/learn-palantype/static-css
Found 1 files

Step 2/4: Looking for hie.yaml files that control setup
Found 1 cradle
  (/home/ruben/code/learn-palantype/static-css/hie.yaml)

Step 3/4: Initializing the IDE
2021-09-22 07:26:30.639003575 [ThreadId 7] DEBUG hls:	Initializing exports map from hiedb
2021-09-22 07:26:30.650772194 [ThreadId 7] DEBUG hls:	Done initializing exports map from hiedb (0)

Step 4/4: Type checking the files
2021-09-22 07:26:30.652362196 [ThreadId 42] DEBUG hls:	hlint:getIdeas:file:NormalizedFilePath "/home/ruben/code/learn-palantype/static-css/app/Main.hs"
2021-09-22 07:26:30.652788241 [ThreadId 44] INFO hls:	Consulting the cradle for "app/Main.hs"
Output from setting up the cradle Cradle {cradleRootDir = "/home/ruben/code/learn-palantype/static-css", cradleOptsProg = CradleAction: Cabal}
> Build profile: -w ghc-8.10.7 -O1
> In order, the following will be built (use -v for more details):
>  - static-css-0.1.0.0 (exe:static-css) (first run)
> Preprocessing executable 'static-css' for static-css-0.1.0.0..
2021-09-22 07:26:30.870926057 [ThreadId 44] DEBUG hls:	Session loading result: Right (ComponentOptions {componentOptions = ["-fbuilding-cabal-package","-O0","-outputdir","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-odir","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-hidir","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-stubdir","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-i","-i/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-iapp","-i/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/autogen","-i/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/global-autogen","-I/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/autogen","-I/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/global-autogen","-I/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/static-css-tmp","-optP-include","-optP/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/build/static-css/autogen/cabal_macros.h","-hide-all-packages","-Wmissing-home-modules","-no-user-package-db","-package-db","/home/ruben/.cabal/store/ghc-8.10.7/package.db","-package-db","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/packagedb/ghc-8.10.7","-package-db","/home/ruben/.cache/hie-bios/dist-static-css-88b960f461f0a2c687c9058dbbff9e00/build/x86_64-linux/ghc-8.10.7/static-css-0.1.0.0/x/static-css/package.conf.inplace","-package-id","base-4.14.3.0","-package-id","clay-0.13.3-BXLxqq6WL6W4JAmsL68AjG","-XHaskell2010","app/Main.hs","-hide-all-packages"], componentRoot = "/home/ruben/code/learn-palantype/static-css", componentDependencies = ["static-css.cabal","cabal.project","cabal.project.local"]},"/nix/store/xwqrc3mzrj1jp18pq66bhf2rmbi4qlqh-ghc-8.10.7-with-packages/lib/ghc-8.10.7")
2021-09-22 07:26:30.916814957 [ThreadId 44] INFO hls:	Using interface files cache dir: /home/ruben/.cache/ghcide/main-c6af42609e289df23abb1eacc99445ce7f846a63
2021-09-22 07:26:30.917074459 [ThreadId 44] INFO hls:	Making new HscEnv[main]
2021-09-22 07:26:30.923801205 [ThreadId 44] DEBUG hls:	New Component Cache HscEnvEq: (([],Just HscEnvEq 5),fromList [("/home/ruben/code/learn-palantype/static-css/hie.yaml",Just 2021-09-21 15:39:44.556397565 UTC),("cabal.project",Just 2021-09-21 15:13:45.417548378 UTC),("cabal.project.local",Nothing),("static-css.cabal",Just 2021-09-21 15:13:45.416548381 UTC)])
2021-09-22 07:26:30.924087233 [ThreadId 44] DEBUG hls:	Known files updated: fromList [(TargetFile NormalizedFilePath "/home/ruben/code/learn-palantype/static-css/app/Main.hs",fromList ["/home/ruben/code/learn-palantype/static-css/app/Main.hs"])]
2021-09-22 07:26:30.92452397 [ThreadId 18] DEBUG hls:	Finishing build session(exception: AsyncCancelled)
2021-09-22 07:26:30.924595898 [ThreadId 44] DEBUG hls:	Restarting build session for keys [GhcSessionIO; ] (aborting the previous one took 0.00s)
2021-09-22 07:26:30.925225935 [ThreadId 87] DEBUG hls:	hlint:getIdeas:file:NormalizedFilePath "/home/ruben/code/learn-palantype/static-css/app/Main.hs"
2021-09-22 07:26:31.017324784 [ThreadId 159] INFO hls:	finish: User TypeCheck (took 0.09s)
2021-09-22 07:26:31.029544268 [ThreadId 170] INFO hls:	finish: GetHie (took 0.01s)
2021-09-22 07:26:31.02974726 [ThreadId 171] INFO hls:	finish: GenerateCore (took 0.00s)

Completed (1 file worked, 0 files failed)
```

**Comment by fendor** [comment]
Well, that looks correct. Maybe some logs from hls within your emacs process? E.g. the logs hls produces while running in the editor? Maybe there is some emacs configuration that is picking the wrong directory?

**Comment by rubenmoor** [comment]
```
Found "/home/ruben/code/learn-palantype/hie.yaml" for "/home/ruben/code/learn-palantype/a"
Run entered for haskell-language-server-wrapper(haskell-language-server-wrapper) Version 1.2.0.0 x86_64 ghc-8.10.7
Current directory: /home/ruben/code/learn-palantype
Operating system: linux
Arguments: ["--lsp","-d","-l","/run/user/1000/hls.log"]
Cradle directory: /home/ruben/code/learn-palantype
Cradle type: Cabal

Tool versions found on the $PATH
cabal:		3.4.0.0
stack:		2.7.3
ghc:		8.10.7


Consulting the cradle to get project GHC version...
Failed to get project GHC version:CradleError {cradleErrorDependencies = [], cradleErrorExitCode = ExitFailure 1, cradleErrorStderr = ["Error when calling cabal --builddir=/home/ruben/.cache/hie-bios/dist-learn-palantype-88327e4bda13069c64fb8dd0be6d913a v2-exec --with-compiler /home/ruben/.cache/hie-bios/wrapper-13a09b18ea883dd377d59db5e821a86b ghc -v0 -- --numeric-version","","cabal: Could not resolve dependencies:\n[__0] trying: backend-0.1 (user goal)\n[__1] trying: servant-snap-0.9.0 (dependency of backend)\n[__2] next goal: base (dependency of backend)\n[__2] rejecting: base-4.14.3.0/installed-4.14.3.0 (conflict: servant-snap =>\nbase>=4.7 && <4.14)\n[__2] skipping: base-4.14.1.0, base-4.14.0.0 (has the same characteristics\nthat caused the previous version to fail: excluded by constraint '>=4.7 &&\n<4.14' from 'servant-snap')\n[__2] rejecting: base-4.13.0.0, base-4.12.0.0, base-4.11.1.0, base-4.11.0.0,\nbase-4.10.1.0, base-4.10.0.0, base-4.9.1.0, base-4.9.0.0, base-4.8.2.0,\nbase-4.8.1.0, base-4.8.0.0, base-4.7.0.2, base-4.7.0.1, base-4.7.0.0,\nbase-4.6.0.1, base-4.6.0.0, base-4.5.1.0, base-4.5.0.0, base-4.4.1.0,\nbase-4.4.0.0, base-4.3.1.0, base-4.3.0.0, base-4.2.0.2, base-4.2.0.1,\nbase-4.2.0.0, base-4.1.0.0, base-4.0.0.0, base-3.0.3.2, base-3.0.3.1\n(constraint from non-upgradeable package requires installed instance)\n[__2] fail (backjumping, conflict set: backend, base, servant-snap)\nAfter searching the rest of the dependency tree exhaustively, these were the\ngoals I've had most trouble fulfilling: base, servant-snap, backend\n\n"]}

Process lsp-haskell stderr finished
```

**Comment by rubenmoor** [comment]
I suspected that it has to do with the way the project root is determined by emacs. As far as I know this is done by proejctile. Then again, a hie.yaml in one of the folders should define a project root, too, or shouldn't it? And could it be that a wrong project root is cached somewhere?

**Comment by rubenmoor** [comment]
Alright, I fixed it.

Of course the root of the problem is lsp-mode or lsp-haskell. Or simply a user error on my part.

https://github.com/emacs-lsp/lsp-haskell/issues/51#issuecomment-619294352

This solved it for me. When no project root is defined, lsp-mode determines it interactively. I did this for the the project in the root folder and thus implicitly set the wrong project root for the subfolder.

By removing the file `~/.emacs.d/.lsp-session-v1` I got to determine the project root anew. By opening the subproject first, I got it set up all correctly now.
